native_client/deepspeech.cc

```diff
@@ -429,7 +480,37 @@ ModelState::decode(vector<float>& logits)
     inputs.data(), n_frames, num_classes, *alphabet, beam_width,
     cutoff_prob, cutoff_top_n, scorer);
 
-  return strdup(alphabet->LabelsToString(out[0].tokens).c_str());
+  return out;
+}
+
+Metadata* ModelState::decode_metadata(vector<float>& logits) 
+{
+  vector<Output> out = decode_raw(logits);
+
+  Metadata* metadata = (Metadata*)malloc(sizeof (Metadata));
+  metadata->num_items = out[0].tokens.size();
+  metadata->items = (MetadataItem*)malloc(sizeof(MetadataItem) * metadata->num_items);
+
+  // Loop through each character
+  for (int i = 0; i < out[0].tokens.size(); ++i) {
+    char* character = (char*)alphabet->StringFromLabel(out[0].tokens[i]).c_str();
+
+    // Note: 1 timestep = 20ms
+    // Timesteps are from the position of the highest letter probability so we 
+    // offset them back by 3ms to account for this
```

Comment:
Maybe move the 3ms heuristic to the client code?

Reply:
So the client should only work for English, German, French... and not Mandarin? That doesn't seem like an idea goal.

Reply:
By default. Until we have a Mandarin model we won't even know what the timings look like, if they need a similar adjustment, how people use timing info for Mandarin, etc. I'm just trying to avoid designing ahead of time for unknown requirements. If, say, Firefox integrates DeepSpeech, they're not going to be using `client.cc`, they'll have to write their own code that targets the API directly, and then decide how to use the timings.

I think as long as the API changes don't artificially limit users, the client can have stuff like the 3ms English heuristic for now, as a form of minimum viable product.

Reply:
Why not have no-offset? Then we are not favoring any language.

Reply:
Well, my only reason is because having the offset is useful for the one language we have a model for right now. The current client timing code is useless for Mandarin, since the concept of splitting into words on spaces doesn't apply there, so we wouldn't use the 3ms offset anyway if it was in the client.

Reply:
I'm coming round to Kelly's point of view. 3ms is a value that I found to be good on my specific test files, all of which are English. It improved the start points of several words in the test files but some were still partly cut off. So it's a nudge in the right direction but not something that will fix every detected word. If we use 3ms in the official clients people might get confused and think it's some kind of official, absolute value.

For my own uses I don't actually need that level of accuracy (I probably only need it accurate to the second), so maybe it's best to have no offset and clients can offset if they need it to be more exact.

It's also worth noting that I only know it was off because I specifically looked at the waveform in an audio editor. For a lot of the words in my test files, it's hard to tell by ear if you just play them back.

Reply:
What's the final verdict on this? 3ms offset in the client or no offset at all?

Reply:
I'll defer to @kdavis-mozilla on this, I'm OK with both options.